README.md
# firtz


*firtz podcast publisher*  
*Version 2.1*

## Über

Du hast offensichtlich den Drang, entweder diesen Kram hier auszuprobieren, oder einen Podcast aus der Taufe zu heben.
Was auch immer Dich dazu bewogen hat, firtz auszuprobieren: Gute Idee :)

Um einen Podcast zu veröffentlichen, reichen im Grunde zwei Dinge: Audiodateien im Web und irgendwo eine XML-Datei, die den Feed beschreibt. Alles andere ist Beiwerk, das mehr oder minder nötig und arbeitsintensiv ist.

Warum aber an vielen Stellen so viel Arbeit machen, so viel Ballast herumschleppen? Wieso ein Wordpress, wenn man alle zwei Monate mal was veröffentlicht? Wieso eine SQL-Datenbank für fünf Artikel in einem Jahr? Wieso sich den Kopf zerbrechen über die Installation diverser Plugins, wenn es ein paar KByte XML tun?

Es gibt aber noch einen zweiten Aspekt. Während ich täglich in der [Hörsuppe](http://hoersuppe.de) über deutschsprachige Podcasts blogge und podcaste, sehe ich mich immer wieder mit Podcasts konfrontiert, die viel Mühe und Arbeit in das gesprochene Wort investieren, dann aber an den technischen Hürden der Veröffentlichung, dem Publishing scheitern.

Das resultiert in kaputten Feeds, Feeds, in denen wichtige Metadaten über die Audiodateien hinaus fehlen und es dem Hörer zwar nicht unmöglich, aber unbequem machen, den Feed zu verfolgen.

Darüber hinaus erleichtert es mir mit der Hörsuppe die Arbeit enorm, wenn die Metadaten stimmen :)

Nun ist Reden Silber, Handeln hingegen Gold. Der firtz soll - wo möglich - dabei helfen, Podcasts den Druck zu nehmen, sich um technische Details zu kümmern.

Deshalb also firtz. Die 0.1 war das Ergebnis von knapp einskommafünf Tagen "Arbeit", die ich eigentlich nur gemacht habe, um mal wieder irgendwas zusammen zu schrauben, das halbwegs funktioniert.

Wie funktioniert das aber nun? Über Konfigurationsdateien. Eine für den Feed mit allgemeinen Daten und einer Datei für jede Episode mit den Details eben jener.

Wer es ganz einfach haben möchte, verbündet sich mit auphonic und überlässt alle Daten, die dort anfallen dem firtz, der dann daraus einen oder mehrere Feeds erzeugt. Einfacher wird's nur noch, wenn Auphonic selbst sowas anbietet ;)


## Voraussetzungen

Aber fangen wir vorne an. Was brauchst Du?

Du brauchst irgendwo Platz im Web. Das ist dann doch noch nötig. Dieser Platz im Web muss im Grunde nur eine Bedingung erfüllen: **php ab Version 5.4.0** muss ausgeführt werden können.

Datenbanken und anderen Kram braucht es nicht, allerdings muss der ausführende Nutzer des Webservers in der Lage, Verzeichnisse zu erzeugen und Dateien in diese hineinzuschreiben.

Sollte der firtz in einem Unterverzeichnis der genutzten Domain betrieben werden, musst Du noch die .htaccess korrigieren und die
RewriteBase von

`RewriteBase /`

auf

`RewriteBase /UNTERORDNER`

ändern.

Es gibt übrigens hier und dort Probleme mit dem Schreiben in das temporäre Verzeichnis. Je nach Webhoster kann es nötig und hilfreich sein, den Handler für php auf (fast-)cgi zu stellen. In der .htaccess geht das z.B. mit dieser Zeile:  

`AddHandler php5-cgi .php`

Das sollte allerdings wirklich nur passieren, wenn der firtz diesbezüglich Fehlermeldungen auswirft.

## Los geht's

Du besorgst Dir das [firtz-Archiv](https://github.com/eazyliving/firtz/), entpackst es in den Ordner, in dem Du das auf dem Webserver haben willst und bewegst Dich darauf hin in den Unterordner `feeds/`.

In diesem findest Du ein Demo-Verzeichnis. Der Name dieses Ordners, ist der zukünftige Name Deines Feeds.

Im Demo-Ordner sind zwei Dateien: `feed.cfg` und `001.epi`. Das nicht syntaktisch zu nennende Prinzip dieser Dateien ist identisch:

`#: ist ein Kommentar nicht den ":" vergessen!`

Attribute werden wie folgt notiert:

`attribute:`

`value`

Achtet bitte darauf, dass die Attribut-Zeile alleine steht. Leerzeilen werden ignoriert, es sei denn, es geht um eine Textfeld wie summary.

An's Ende der Konfiguration kannst Du in einer neuen Zeile ein

`---end---`

setzen, alles dahinter wird ignoriert. Du kannst da Notizen oder Müll unterbringen ;) Ich werde das hier im Moment noch nicht weiter ausbreiten. Schau in die Dateien und spiel damit rum.

**Wichtig ist, dass alle Konfigurationsdateien UTF8 sind, denn sonst gibt's Müsli im Feed!**

Solltest Du [Auphonic](http://auphonic.com/) nutzen, bist Du ganz fein raus. Aktiviere in den Ausgangsdateien dort die Production-Description, sorge dafür, dass firtz sie im Dateisystem findet und schon erstellen sich die Folgen fast von selbst anhand der von Auphonic weitergereichten Metadaten. Weitere Informationen dazu finden sich in der weiterführenden Dokumentation.

Nehmen wir an, Du hast das soweit fertig. Wie erreichst du den Feed? Angenommen der URL zu Deinem Webserver wäre `http://tollerneuerpodcast.de/`. Der Feed heißt ja noch demo, also wäre der URL zum RSS2-Feed:

`http://tollerneuerpodcast.de/demo`

Wenn Du mehrere Audioformate nutzt, kannst Du auch z.B.
`http://tollerneuerpodcast.de/demo/mp3` nutzen.

Eine Webseite gibt's als Bonus. Die kannst Du für diesen Feed so erreichen: `http://tollerneuerpodcast.de/demo/show`

Für jede Episode gibt's dann einen eigenen [Podlove Webplayer](https://github.com/gerritvanaaken/podlove-web-player) und [flattr](http://flattr.com)-Buttons.

Einzelne Episoden sind auch verlinkbar, indem an den Pfad der Slug der Episode (praktisch der Dateiname ohne Endung) angehangen wird: `http://tollerneuerpodcst.de/demo/show/001`.

## Kontakt

- [@firtz](https://twitter.com/the_firtz) (project)
- [@eazy](https://twitter.com/eazyliving)
- [@mccouman](https://twitter.com/mccouman)


## Links

- [fatfree framework](https://github.com/bcosca/fatfree)
- [QuorX II](https://github.com/Firtz-Designs/QuorX-II)
- [Podlove Webplayer](https://github.com/gerritvanaaken/podlove-web-player)
- [Ein guter Artikel von Hirnbloggade zur firtz Installation](http://www.hirnbloggade.de/2013/04/26/der-firtz-und-ich/)


CHANGELOG.md
Changelog:

RC 2.1 Beta 2
- fixed site.html
- fixed extension (description as JSON output)
- fix share json
- fix feed class
- remove pwp folder (obsolet!)

RC 2.1 firtz + podlove webplayer 4  
- remove search extension and include in firtz (redirect)
- new extension "webplayer" with podlove webplayer version 4 (standrad include)
- fixed opus file and other minimalistics
- create share rendering function in feed.php
- new endpoint for serach requests
- new endpoint to send sharing request
- fixed weblayer css and positions

RC 2.0

1.5 so far:
cleaned structure:
- lib, pwp, js, dict and classes moved to src-folder

changed way templating works:
- main template-folder contains default template (now quorx!)
- additional folders inside template-folder can contain additional files, overriding files inside default folder
- feed.cfg needs template: attribute
- feed.cfg can contain template-vars (go to @templatevars)
- template.cfg inside template-folder can contain default template-vars

1.3 update to podlove webplayer v2.0.13 and support images in mp4chaps

1.2 enabled rfc5005 inside standard feed (until now only by special urls...)
- fixed onother pile of smelly bugs. Take some care though...

1.1 another tons of bugfixes, templates per feed and podlove webplayer 2.0.11
- updated to fatfree 3.0.9, added adnthreads for episode comments

1.0 tons of bugs fixed. podlove webplayer version 2.0.7 integrated

0.13 changed path to pages (now /show/page/@page from /page/@page) and pagers (now /show/pager/@pagenum from /show/page/@pagenum)
- added subdirs for /templates/pages: create dropdowns. only one level
- changed to podlove web player 2.0.5

0.12 fixed bug in markdown parsing (thanks bcosca)
- first (stable? haha) cloning support. call index.php from commandline

0.11 fixed pwp integration bug

0.10 changed to podlove webplayer v2.0.4

0.9 added bitlove support; reintroduced dropdown menus; articles now marked up by markdown

0.8 tons of fixes, typos and stuff. content-extensions.

0.7 autodetect mediafiles with mediabaseurl/path and formats:attribute

0.6 finished first draft of documentation. see doc/firtz.md

0.5 added subtemplate support for head/header/episode/footer
- fixed a bug, in case episodes didn't provide all configured media

0.4 added local auphonic support, yet work in progress
- singlefeed pages redirect to web page on / instead if listing one feed on frontpage

0.3 added first draft of extension class. atm only extends output templates (example: atom-extension. look into ext/atom/)
- changed website URL from /show/$feed to $feed/show to keep consistancy

0.2 major and minor fuckups fixed. thanks to all contributors!

0.1 let the world take a look


index.php
<?php
# RC 2.1 Beta 2
ini_set('auto_detect_line_endings', true);

$main = require('src/lib/base.php');

$main->set('FEEDDIR', './feeds');
$main->set('EXTDIR', './ext');
$main->set('BITMASK', ENT_NOQUOTES | ENT_XML1);
$main->set('UI', '');
$main->set('templatepath', 'templates/default/');
$main->set('version', 2);
$main->set('revision', 1);
$main->set('generator', 'firtz podcast publisher v' . $main->get('version') . "." . $main->get('revision'));
$main->set('pager', '');
$main->set('BASEURL', "http://" . str_replace("/", "", $main->get('HOST')) . dirname($_SERVER['SCRIPT_NAME']));
$main->set('BASEPATH', $_SERVER['DOCUMENT_ROOT']);
$main->set('singlepage', false);
$main->set('showpage', false);
$main->set('AUTOLOAD', 'src/classes/');
$main->set('CDURATION', 3600);
$main->set('page', 0);
$main->set('DEBUG', 2);
$main->set('epi', '');
$main->set('og', array());
$main->set('clonemode', false);
$main->set('extxml', '');
$main->set('extvars', array());
$main->set('exthtml', '');
$main->set('LOCALES', 'src/dict/');
$main->set('rfc5005', '');
$main->set('audio', '');
$main->set('search', '');

if (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] != "") $main->set('scheme', 'https'); else $main->set('scheme', 'http');

$main->set('firtzattr_default', array('feedalias', 'baseurlredirect'));
$main->set('feedattr_default', array('title', 'description', 'formats', 'flattrid', 'author', 'summary', 'image', 'keywords', 'category', 'email', 'language', 'explicit', 'itunes', 'auphonic-path', 'auphonic-glob', 'auphonic-url', 'auphonic-mode', 'twitter', 'itunesblock', 'mediabaseurl', 'mediabasepath', 'redirect', 'bitlove', 'cloneurl', 'clonepath', 'licenseurl', 'licensename', 'licenseimage', 'rfc5005', 'baseurl', 'feedalias', 'articles-per-page', 'template', 'templatevars'));
$main->set('itemattr', array('title', 'description', 'link', 'guid', 'article', 'payment', 'chapters', 'enclosure', 'duration', 'keywords', 'image', 'date', 'noaudio', 'location'));
$main->set('extattr', array('slug', 'template', 'arguments', 'prio', 'script', 'type', 'vars', 'episode-vars', 'feed-vars'));
$main->set('mimetypes',
    array(
        'mp3' => 'audio/mpeg',
        'torrent' => 'application/x-bittorrent',
        'mpg' => 'video/mpeg',
        'm4a' => 'audio/mp4',
        'm4v' => 'video/mp4',
        'oga' => 'audio/ogg',
        'ogg' => 'audio/ogg',
        'ogv' => 'video/ogg',
        'webma' => 'audio/webm',
        'webm' => 'video/webm',
        'flac' => 'audio/flac',
        'opus' => 'audio/opus', #'audio/ogg;codecs=opus',
        'mka' => 'audio/x-matroska',
        'mkv' => 'video/x-matroska',
        'pdf' => 'application/pdf',
        'epub' => 'application/epub+zip',
        'png' => 'image/png',
        'jpg' => 'image/jpeg',
        'mobi' => 'application/x-mobipocket-ebook'
    )
);

$firtz = new firtz($main);
$firtz->loadAllTheExtensions($main);
$main->set('firtz', $firtz);
$feeds = array();

foreach (glob($main->get('FEEDDIR') . '/*', GLOB_ONLYDIR) as $dir) {
    if (substr(basename($dir), 0, 1) != "_") $feeds[] = basename($dir);
}

$main->set('feeds', $feeds);

function firtzConvertAmp($content)
{
    echo preg_replace('/&([^#])(?![a-z1-4]{1,8};)/i', '&#038;$1', $content);
}

function sortByPubDate($a, $b)
{
    if (strtotime($a->item['pubDate']) == strtotime($b->item['pubDate'])) return ($a->item['slug'] < $b->item['slug']);
    return (strtotime($a->item['pubDate']) < strtotime($b->item['pubDate']));

}

function strip_article($text, $allowed_tags = array())
{

    $rtext = preg_replace_callback('/<\/?([^>\s]+)[^>]*>/i', function ($matches) use (&$allowed_tags) {
        return in_array(strtolower($matches[1]), $allowed_tags) ? $matches[0] : '';
    }, $text);
    return $rtext;
}

function firtz_markdown($text)
{
    global $main;
    $f = $main->get('firtz');
    return $f->markdown->text($text);
}

foreach ($firtz->attr['feedalias'] as $alias) {

    $main->route('GET|HEAD ' . $alias['route'],

        function ($main, $params) use ($alias) {

            header('HTTP/1.1 301 Moved Permanently');
            header('Location: /' . $alias['feed'] . '/' . $alias['format']);
            die();

        }
    );

}


foreach ($firtz->extensions as $slug => $extension) {
    if ($extension->type != 'output') continue;
    $slug = $extension->slug;
    $extension->init();

    $main->route("GET|HEAD /@feed/$slug/*",
        function ($main, $params) use ($slug) {

            $firtz = $main->get('firtz');
            $extension = $firtz->extensions[$slug];

            $arguments = array();
            $arguments_ext = $extension->arguments;
            $arguments_get = explode("/", $params[2]);

            foreach ($arguments_get as $key => $val) {
                if (isset($arguments_ext[$key])) {
                    $argname = $arguments_ext[$key];
                    $arguments[$argname] = $val;
                    $main->set($argname, $val);
                } else {
                    $main->set($argname, '');
                }

            }

            $extension->arguments = $arguments;

            $feedslug = $params['feed'];
            if (!in_array($params['feed'], $main->get('feeds'))) $main->error(404);

            $BASEPATH = $main->get('FEEDDIR') . '/' . $params['feed'];
            $FEEDCONFIG = $BASEPATH . '/feed.cfg';

            $feed = new feed($main, $feedslug, $FEEDCONFIG);
            $feed->findEpisodes();
            $feed->loadEpisodes();
            $feed->runExt($main, $extension);
        }, $main->get('CDURATION')
    );

}

foreach ($firtz->extensions as $slug => $extension) {

    if ($extension->type != 'output') continue;

    $slug = $extension->slug;
    $main->route("GET|HEAD /@feed/$slug",
        function ($main, $params) use ($slug) {
            $firtz = $main->get('firtz');

            $extension = $firtz->extensions[$slug];

            $arguments = array();
            $arguments_ext = $extension->arguments;

            foreach ($arguments_ext as $argname) {
                $arguments[$argname] = "";
                $main->set($argname, "");
            }

            $extension->arguments = $arguments;

            $feedslug = $params['feed'];
            if (!in_array($params['feed'], $main->get('feeds'))) $main->error(404);

            $BASEPATH = $main->get('FEEDDIR') . '/' . $params['feed'];
            $FEEDCONFIG = $BASEPATH . '/feed.cfg';

            $feed = new feed($main, $feedslug, $FEEDCONFIG);
            $feed->findEpisodes();
            $feed->loadEpisodes();
            $feed->runExt($main, $extension);
        }, $main->get('CDURATION')
    );

}


/*
	direct call for a specified feed/audio-combination
	
*/

$main->route('GET|HEAD /@feed/@audio',
    function ($main, $params) {

        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';

        $feed = new feed($main, $slug, $FEEDCONFIG);

        if ($feed->attr['redirect'] != "") {
            header('HTTP/1.1 301 Moved Permanently');
            header('Location: ' . $feed->attr['redirect']);
            die();
        }

        $main->set('audio', $params['audio']);

        $feed->findEpisodes();
        $feed->loadEpisodes();
        if ($feed->attr['rfc5005'] == "on") {
            $main->set('rfc5005', 'on');
            $main->set('maxpage', ceil(sizeof($feed->episodes) / 10));
            $feed->episodes = array_slice($feed->episodes, 0, 10);
        }

        $feed->renderRSS2($params['audio']);

    }, $main->get('CDURATION')
);

$main->route('GET|HEAD /@feed/map',
    function ($main, $params) {

        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';

        $feed = new feed($main, $slug, $FEEDCONFIG);

        $feed->findEpisodes();
        $feed->loadEpisodes();

        $feed->renderMap();

    }, $main->get('CDURATION')
);


/*
	get the main feed (audioformat according to formats: attribute in feed.cfg
	
*/

$main->route('GET|HEAD /@feed',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';

        $feed = new feed($main, $slug, $FEEDCONFIG);

        if ($feed->attr['redirect'] != "") {
            header('HTTP/1.1 301 Moved Permanently');
            header('Location: ' . $feed->attr['redirect']);
            die();
        }

        $feed->findEpisodes();
        $feed->loadEpisodes();

        if ($feed->attr['rfc5005'] == "on") {
            $main->set('rfc5005', 'on');
            $main->set('maxpage', ceil(sizeof($feed->episodes) / 10));
            $feed->episodes = array_slice($feed->episodes, 0, 10);
        }


        $feed->renderRSS2();
    }, $main->get('CDURATION')
);

$main->route('GET|HEAD /@feed/page/@page',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';

        $feed = new feed($main, $slug, $FEEDCONFIG);

        if ($feed->attr['redirect'] != "") {
            header('HTTP/1.1 301 Moved Permanently');
            header('Location: ' . $feed->attr['redirect']);
            die();
        }

        $main->set('rfc5005', 'on');

        $feed->findEpisodes();
        $feed->loadEpisodes();

        $main->set('page', ltrim($params['page'], '0'));
        $main->set('maxpage', ceil(sizeof($feed->episodes) / 10));

        if ($main->get('page') == 'first' || $main->get('page') == 'current') $main->set('page', 1);
        if ($main->get('page') == 'last') $main->set('page', $main->get('maxpage'));

        $feed->episodes = array_slice($feed->episodes, ($main->get('page') - 1) * 10, 10);

        $feed->renderRSS2();

    }, $main->get('CDURATION')
);

$main->route('GET|HEAD /@feed/@audio/page/@page',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';

        $feed = new feed($main, $slug, $FEEDCONFIG);

        if ($feed->attr['redirect'] != "") {
            header('HTTP/1.1 301 Moved Permanently');
            header('Location: ' . $feed->attr['redirect']);
            die();
        }

        $main->set('rfc5005', 'on');

        $feed->findEpisodes();
        $feed->loadEpisodes();

        $main->set('page', ltrim($params['page'], '0'));
        $main->set('maxpage', ceil(sizeof($feed->episodes) / 10));

        if ($main->get('page') == 'first' || $main->get('page') == 'current') $main->set('page', 1);
        if ($main->get('page') == 'last') $main->set('page', $main->get('maxpage'));

        $feed->episodes = array_slice($feed->episodes, ($main->get('page') - 1) * 10, 10);
        $main->set('audio', $params['audio']);

        $feed->renderRSS2($params['audio']);

    }, $main->get('CDURATION')
);


/*
	web page mode, complete page
	paging mode comes next

*/

$main->route('GET|HEAD /@feed/show',
    function ($main, $params) {

        $slug = $params['feed'];

        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';

        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();
        $feed->loadEpisodes();
        $main->set('page', 1);
        $main->set('maxpage', ceil(sizeof($feed->episodes) / $feed->attr['articles-per-page']));
        $feed->episodes = array_slice($feed->episodes, 0, $feed->attr['articles-per-page']);

        $feed->renderHTML();

    }, $main->get('CDURATION')
);

/*
	web page mode, single page for episode
*/

$main->route('GET|HEAD /@feed/show/@epi',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';
        $main->set('singlepage', true);
        $main->set('epi', $params['epi']);
        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();
        $feed->loadEpisodes($params['epi']);
        $feed->renderHTML();
    }, $main->get('CDURATION')

);

/*
 * Share: podlove webplayer 4 | /share =>html
 */
$main->route('GET /share',
    function ($main) {
        echo Template::instance()->render('share.html');
    }, $main->get('CDURATION')
);

/**
 * Share Episode
 */
$main->route('GET|HEAD /@feed/share/@epi',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';
        $main->set('epi', $params['epi']);
        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();
        $feed->loadEpisodes($params['epi']);
        $feed->renderShare($params['epi']);

    }, $main->get('CDURATION')
);

$main->route('GET|HEAD /@feed/show/search',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $main->reroute('/' . $slug . '/show');

    }, $main->get('CDURATION')
);

$main->route('GET|HEAD /@feed/show/search/@tag',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';
        if ($params['tag'] == "") $main->reroute('/' . $slug . '/show');
        $main->set('search', $params['tag']);
        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();
        $feed->loadEpisodes();
        $feed->renderHTML();
    }, $main->get('CDURATION')
);

$main->route('GET|HEAD /@feed/show/search/@tag/@audio',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';
        if ($params['tag'] == "") $main->reroute('/' . $slug . '/show');
        $main->set('search', $params['tag']);
        $feed = new feed($main, $slug, $FEEDCONFIG);

        $main->set('audio', $params['audio']);

        $feed->findEpisodes();
        $feed->loadEpisodes();
        if ($feed->attr['rfc5005'] == "on") {
            $main->set('rfc5005', 'on');
            $main->set('maxpage', ceil(sizeof($feed->episodes) / 10));
            $feed->episodes = array_slice($feed->episodes, 0, 10);
        }


        $feed->renderRSS2($params['audio']);

    }, $main->get('CDURATION')
);

/*
 * Search: redirect
 */
$main->route('GET /search',
    function ($main) {
        if (isset($_GET["s"]) && !empty($_GET["s"]) &&
            isset($_GET["r"]) && !empty($_GET["r"])
        ) {
            header('location: ' . @$_GET["r"] . '/show/search/' . @$_GET["s"] . '');
        }
    }
);


$main->route('GET|HEAD /@feed/show/@epi/@ignore',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';
        $main->set('singlepage', true);
        $main->set('epi', $params['epi']);
        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();
        $feed->loadEpisodes($params['epi']);
        $feed->renderHTML();
    }, $main->get('CDURATION')
);

/*
	web page mode, pageing 3 shows
*/

$main->route('GET|HEAD /@feed/show/pager/@pagenum',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);
        $pagenum = ltrim($params['pagenum'], '0');
        if (!is_numeric($pagenum)) $pagenum = 1;

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';

        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();

        $feed->loadEpisodes();
        $main->set('page', $pagenum);
        $main->set('maxpage', ceil(sizeof($feed->episodes) / $feed->attr['articles-per-page']));

        $feed->episodes = array_slice($feed->episodes, ($pagenum - 1) * $feed->attr['articles-per-page'], $feed->attr['articles-per-page']);
        $feed->renderHTML();
    }, $main->get('CDURATION')
);


/*
	main page without any parameters
	simple list of available podcasts (web page)
	if single feed page, then redirect to web page
   
*/

$main->route('GET /',
    function ($main, $params) {

        $feeds = array();
        $firtz = $main->get('firtz');
        $allfeeds = $main->get('feeds');
        if (sizeof($allfeeds) == 1) $main->reroute('/' . $allfeeds[0] . '/show');

        foreach ($allfeeds as $slug) {


            $FEEDPATH = $main->get('FEEDDIR') . '/' . $slug;
            $FEEDCONFIG = $FEEDPATH . '/feed.cfg';
            $feed = new feed($main, $slug, $FEEDCONFIG);
            $feeds[] = $feed->attr;
            if ('http://' . $main->get('HOST') == $feed->attr['baseurl']) {
                $main->reroute('/' . $slug . '/show');
            }
            $main->set('frontlanguage', substr($feed->attr['language'], 0, 2));
            $main->set('fronttitle', $feed->attr['title']);
            $main->set('frontauthor', $feed->attr['author']);
        }

        $main->set('frontfeeds', $feeds);
        echo Template::instance()->render('front.html');
    }, $main->get('CDURATION')
);

/*
	single page mode with custom content page
	put them im templates/pages/
*/

$main->route('GET|HEAD /@feed/show/page/@page',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';
        $main->set('singlepage', true);
        $main->set('pagename', $params['page']);
        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();
        $feed->loadEpisodes();
        $feed->renderHTML(false, $params['page']);

    }, $main->get('CDURATION')
);

$main->route('GET|HEAD /@feed/show/page/@dir/@page',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';
        $main->set('singlepage', true);
        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();
        $feed->loadEpisodes();
        $feed->renderHTML(false, $params['dir'] . '/' . $params['page']);

    }, $main->get('CDURATION')
);

$main->route('GET|HEAD /@feed/raw/@epi',
    function ($main, $params) {
        $slug = $params['feed'];
        if (!in_array($slug, $main->get('feeds'))) $main->error(404);

        $BASEPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $BASEPATH . '/feed.cfg';
        $main->set('epi', $params['epi']);
        $feed = new feed($main, $slug, $FEEDCONFIG);
        $feed->findEpisodes();
        $feed->loadEpisodes($params['epi']);

        $feed->renderRaw($params['epi']);
    }, $main->get('CDURATION')
);


if (php_sapi_name() == "cli") {

    function dir_recurse_copy($src, $dst)
    {
        $dir = opendir($src);
        @mkdir($dst);
        while (false !== ($file = readdir($dir))) {
            if (($file != '.') && ($file != '..')) {
                if (is_dir($src . '/' . $file)) {
                    dir_recurse_copy($src . '/' . $file, $dst . '/' . $file);
                } else {
                    copy($src . '/' . $file, $dst . '/' . $file);
                }
            }
        }
        closedir($dir);
    }

    $main->set('clonemode', true);
    $main->set('extxml', '.xml');
    $main->set('exthtml', '.html');

    # CLI mode... create static pages
    foreach ($main->get('feeds') as $slug) {

        $FEEDPATH = $main->get('FEEDDIR') . '/' . $slug;
        $FEEDCONFIG = $FEEDPATH . '/feed.cfg';

        $feed = new feed($main, $slug, $FEEDCONFIG);

        if ($feed->attr['cloneurl'] == '' || $feed->attr['clonepath'] == '') continue;
        $DEST = $feed->attr['clonepath'];

        if (!file_exists($DEST)) {
            @mkdir($DEST);
            if (!file_exists($DEST)) {
                echo "could not create destination path $DEST";
                exit;
            }
        }


        if (!is_writable($DEST)) {
            echo "no permissions to write to $DEST!";
            exit;
        }

        if (strpos($feed->attr['sitecss'], '/feeds/') !== false) {
            $origcss = $FEEDPATH . '/' . basename($feed->attr['sitecss']);
            $clonecss = $main->fixslashes($DEST . '/css/' . basename($feed->attr['sitecss']));
            copy($origcss, $clonecss);
            $feed->attr['sitecss'] = $feed->attr['cloneurl'] . '/css/' . basename($feed->attr['sitecss']);
        }

        dir_recurse_copy('js', $DEST . '/js');
        dir_recurse_copy('css', $DEST . '/css');
        dir_recurse_copy('pwp', $DEST . '/pwp');

        $DEST = $main->fixslashes($DEST . '/' . $slug);
        @mkdir($DEST);
        @mkdir($main->fixslashes($DEST . '/show/'));

        $main->set('BASEURL', $feed->attr['cloneurl']);

        $feed->findEpisodes();
        $feed->loadEpisodes();

        foreach ($feed->attr['audioformats'] as $audio) {
            $xml = $feed->renderRSS2($audio, true);
            file_put_contents($DEST . '/' . $audio . '.xml', $xml);
        }

        $main->set('epi', '');
        $html = $feed->renderHTML(true);

        file_put_contents($DEST . '/show/index.html', $html);
        foreach ($feed->real_slugs as $episode_slug) {
            $main->set('epi', $episode_slug);
            $html = $feed->renderHTML(true);
            @mkdir($DEST . '/show/' . $episode_slug);
            file_put_contents($DEST . '/show/' . $episode_slug . '/index.html', $html);
        }

        foreach (glob('templates/pages/*.html') as $page) {
            $html = $feed->renderHTML(true, basename($page, '.html'));
            @mkdir($main->fixslashes($DEST . '/page'));
            file_put_contents($DEST . '/page/' . basename($page), $html);
        }


        $frontfeeds[] = $feed->attr;
        $main->set('frontlanguage', substr($feed->attr['language'], 0, 2));
        $main->set('fronttitle', $feed->attr['title']);
        $main->set('frontauthor', $feed->attr['author']);

    }
    $main->set('frontfeeds', $frontfeeds);
    $front = Template::instance()->render('front.html');
    file_put_contents($feed->attr['clonepath'] . 'index.html', $front);
    exit;
}

$main->run();

?>
